### ISSUE TITLE
Investigate ophyd test failures

### ISSUE THREAD

**Issue by danielballan** [NSLS-II/caproto#328]
It looks like we introduced a regression at some point, only just now noticed in [this ophyd build](https://travis-ci.org/NSLS-II/ophyd/jobs/397632649).

The proximate error is:

```py
dtype = <ChannelType.CHAR: 4>, values = [b'/data/2018/06/28/\x00']
    def python_to_epics(dtype, values, *, byteswap=True, convert_from=None):
        'Convert python builtin values to epics CA'
        # NOTE: ignoring byteswap, storing everything as big-endian
        if dtype == ChannelType.STRING:
            return DbrStringArray(values).tobytes()
    
>       return np.asarray(values).astype(type_map[dtype])
E       ValueError: invalid literal for int() with base 10: '/data/2018/06/28/'
```

Detailed excerpt from build logs:

<details>

```
_________ test_fstiff_plugin[caproto-None-/data/%Y/%m/%d-None-False1] __________
root = None, wpath = '/data/%Y/%m/%d', rpath = None, check_files = False
    @pytest.mark.parametrize('root,wpath,rpath,check_files',
                             ((None, '/data/%Y/%m/%d', None, False),
                              (None, '/data/%Y/%m/%d', None, False),
                              ('/data', '%Y/%m/%d', None, False),
                              ('/data', '/data/%Y/%m/%d', '%Y/%m/%d', False),
                              ('/', '/data/%Y/%m/%d', None, False),
                              ('/tmp/data', '/data/%Y/%m/%d', '%Y/%m/%d', True)
                              ))
    def test_fstiff_plugin(root, wpath, rpath, check_files):
        fs = DummyFS()
        fs2 = DummyFS()
        if check_files:
            fh = pytest.importorskip('databroker.assets.handlers')
    
        class FS_tiff(TIFFPlugin, FileStoreTIFF,
                      FileStoreIterativeWrite):
            pass
    
        class MyDetector(SingleTrigger, SimDetector):
            tiff1 = Cpt(FS_tiff, 'TIFF1:',
                        write_path_template=wpath,
                        read_path_template=rpath,
                        root=root, reg=fs)
        target_root = root or '/'
        det = MyDetector(prefix, name='det')
        det.read_attrs = ['tiff1']
        det.tiff1.read_attrs = []
    
>       det.stage()
ophyd/tests/test_areadetector.py:338: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ophyd/areadetector/trigger_mixins.py:121: in stage
    super().stage()
ophyd/areadetector/base.py:158: in stage
    ret = super().stage(*args, **kwargs)
ophyd/device.py:607: in stage
    device.stage()
ophyd/areadetector/plugins.py:101: in stage
    super().stage()
ophyd/areadetector/base.py:158: in stage
    ret = super().stage(*args, **kwargs)
ophyd/areadetector/filestore_mixins.py:480: in stage
    super().stage()
ophyd/areadetector/filestore_mixins.py:431: in stage
    set_and_wait(self.file_path, write_path)
ophyd/utils/epics_pvs.py:228: in set_and_wait
    signal.put(val)
ophyd/signal.py:899: in put
    self._write_pv.put(value, use_complete=use_complete, **kwargs)
../../../mc/envs/testenv/lib/python3.6/site-packages/caproto/threading/pyepics_compat.py:28: in inner
    return func(self, *args, **kwargs)
../../../mc/envs/testenv/lib/python3.6/site-packages/caproto/threading/pyepics_compat.py:472: in put
    timeout=timeout, notify=notify)
../../../mc/envs/testenv/lib/python3.6/site-packages/caproto/threading/client.py:101: in inner
    return func(self, *args, **kwargs)
../../../mc/envs/testenv/lib/python3.6/site-packages/caproto/threading/client.py:1491: in write
    data_count=data_count)
../../../mc/envs/testenv/lib/python3.6/site-packages/caproto/_circuit.py:655: in write
    metadata=metadata)
../../../mc/envs/testenv/lib/python3.6/site-packages/caproto/_commands.py:1126: in __init__
    size, *buffers = data_payload(data, metadata, data_type, data_count)
../../../mc/envs/testenv/lib/python3.6/site-packages/caproto/_commands.py:258: in data_payload
    native_type(data_type), data, byteswap=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dtype = <ChannelType.CHAR: 4>, values = [b'/data/2018/06/28/\x00']
    def python_to_epics(dtype, values, *, byteswap=True, convert_from=None):
        'Convert python builtin values to epics CA'
        # NOTE: ignoring byteswap, storing everything as big-endian
        if dtype == ChannelType.STRING:
            return DbrStringArray(values).tobytes()
    
>       return np.asarray(values).astype(type_map[dtype])
E       ValueError: invalid literal for int() with base 10: '/data/2018/06/28/'
../../../mc/envs/testenv/lib/python3.6/site-packages/caproto/_numpy_backend.py:59: ValueError
```

</details>

**Comment by danielballan** [comment]
Seems like the problem was introduced between [this passing build at 12 June 16:39:00](https://travis-ci.org/NSLS-II/ophyd/builds/391393985) and [this failing build at 14 June 17:26:10](https://travis-ci.org/NSLS-II/ophyd/builds/392414191). Many builds of `master` before 12 June also failed but, for example, [the immediately preceding one](https://travis-ci.org/NSLS-II/ophyd/builds/388735912) failed on an unrelated error.

**Comment by danielballan** [comment]
Oh, that was a busy two days of commit activity on the `master` branch:

```bash
$ git log --since=2018-06-12T16:00:00 --until=2018-06-14T18:00:00 --oneline | wc -l
67
```

Via `git bisect`:
```
76a27354ec602766e7ff725e55f2ae654ce99761 is the first bad commit
commit 76a27354ec602766e7ff725e55f2ae654ce99761
Author: danielballan <daniel.b.allan@gmail.com>
Date:   Thu Jun 14 00:22:52 2018 -0500

    ENH: Accept scalar or str/bytes in write.
    
    The sync client already does this. (Checked interactively.)
    
    This PR makes the threading client do the same and adds a test.

:040000 040000 a03d22b0ce301da664bd88f3c9a1e62affb4fd5c 997c0de2099b84bd4e219eeb38643ca4bdafac86 M	caproto
```

**Comment by tacaswell** [comment]
I guess `data_count=0` for char arrays makes sense if you also understand 'native data count' to be the maximum, then this gets the IOC out of the business of understanding where the null byte is and having to do any sort of string parsing.

**Comment by danielballan** [comment]
For the record, it looks like caproto is actually correct here. The issue does not seem to be related to `data_count` after all... even though there are some lurking `data_count` questions unrelated to this issue -- see https://github.com/NSLS-II/caproto/pull/261.

**Comment by danielballan** [comment]
Closed by #351
